File: April 2022/April_05.cpp
class Solution {
public:
    int find(vector<pair<int, int> > &arr, int val){
        int ans = arr.size(), lo = 0, hi = ans - 1;
        while(lo <= hi){
            int mid = (lo + hi) / 2;
            if(arr[mid].first < val)
                lo = mid + 1;
            else{
                ans = mid;
                hi = mid - 1;
            }
        }
        return ans;
    }
    
    int maxArea(vector<int>& height) {
        int n = height.size();
        vector<int> left(n), right(n);
        
        vector<pair<int, int> > arr;
        for(int i = 0; i < n; i++){
            int idx = find(arr, height[i]);
            if(idx == arr.size())
                arr.push_back(make_pair(height[i], i));
            left[i] = arr[idx].second;
        }
        
        arr.clear();
        for(int i = n - 1; i >= 0; i--){
            int idx = find(arr, height[i]);
            if(idx == arr.size())
                arr.push_back(make_pair(height[i], i));
            right[i] = arr[idx].second;
        }
        
        int area = 0;
        for(int i = 0; i < n; i++)
            area = max(area, (right[i] - left[i]) * height[i]);
        
        return area;
    }
};

File: April 2022/April_05.java
class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        int[] left = new int[n], right = new int[n];
        
        List<int[]> list = new ArrayList<>();
        for(int i = 0; i < n; i++){
            int idx = find(list, height[i]);
            if(idx == list.size())
                list.add(new int[]{height[i], i});
            left[i] = list.get(idx)[1];
        }
        
        list.clear();
        for(int i = n - 1; i >= 0; i--){
            int idx = find(list, height[i]);
            if(idx == list.size())
                list.add(new int[]{height[i], i});
            right[i] = list.get(idx)[1];
        }
        
        int area = 0;
        for(int i = 0; i < n; i++)
            area = Math.max(area, height[i] * (right[i] - left[i]));
        
        return area;
    }
    
    private int find(List<int[]> list, int val){
        int ans = list.size(), lo = 0, hi = ans - 1;
        while(lo <= hi){
            int mid = (lo + hi) / 2;
            if(list.get(mid)[0] < val)
                lo = mid + 1;
            else{
                ans = mid;
                hi = mid - 1;
            }
        }
        return ans;
    }
}
